### PR TITLE
Build project

### PR DESCRIPTION
**Pull request by prolajumokeoni**
# Enumerable method
## For this project:
- Followed the git-flow guidelines
- I implemented my own version of ruby's enumerable methods
- We used the yield and #call keywords as required by the instructions to come up with our own versions of these enumerable methods.
- Disabled some style lints so the code  could pass the linter errors

### CODE REVIEW

**Comment by mimipeshy** [comment]
## STATUS : CHANGES REQUESTED 🔴 ♻️ 

Hello team @prolajumokeoni 👋🏾 👋🏾 , this is a really strong first submission. Keep it up 👏🏾 

#### Highlights
- You have followed the correct git-flow ✔️ 
- Your linter checks are passing ✔️ 
- Most of your methods are working ✔️ 

#### To improve
Kindly check the comment section and let's work on the following failing methods  ⬇️ 

That's all. 😄 , kindy submit another review when you are done

Regards
Peris Ndanu
_If you have any questions, kindly leave a comment here and tag my name @mimipeshy and I will be glad to help_

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    my_array = []
+    my_each do |item|
+      my_array.push(item) if yield(i)
+    end
+    my_array
+  end
```
Comment (mimipeshy) [inline enumerables.rb]: - [x] On line 24, you are yielding a variable `i` which you have not defined. Did you mean to yield `item`? Kindly fix this

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+      my_each do |item|
+        match += 1 if yield item
+      end
+
+    else
+      match = length
+    end
+
+    match
+  end
```
Comment (mimipeshy) [inline enumerables.rb]: - [x] enumerables #my_count should return the number of items in enum through enumeration::range. You have an undefined variable named `length`. Kindly fix this

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+      my_each do |item|
+        new_arr.push(proc.call(item))
+      end
+    else
+      my_each do |item|
+        new_arr.push(yield item)
+      end
+    end
+    new_arr
+  end
```
Comment (mimipeshy) [inline enumerables.rb]: - [x] enumerables #my_map should return an Enumerator if no block is given

File: enumerables.rb (hunk 1 above)
Comment (prolajumokeoni) [reply enumerables.rb]: Thank you for your review, I would implement the change and ensure all methods are working @mimipeshy

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+      acc_value = nil
+    end
+    if !block_given? && !symbol.nil?
+      my_each { |i| acc_value = acc_value.nil? ? i : acc_value.send(symbol, i) }
+    else
+      my_each { |i| acc_value = acc_value.nil? ? i : yield(acc_value, i) }
+    end
+    acc_value
+  end
+
```
Comment (mimipeshy) [inline enumerables.rb]: - [x] enumerables #my_inject when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
     Failure/Error: actual = range.my_inject(4) { |prod, n| prod * n }
     
- [x] enumerables #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method
     Failure/Error: expect(array.my_inject(:+)).to eq array.inject(:+)

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  # rubocop:enable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+
+  def multiply_els(arr)
+    arr.my_inject(1, '*')
+  end
```
Comment (mimipeshy) [inline enumerables.rb]: - [x] Since #mutliply_els is not an actual ruby enumerable, kindly define the method outside the enumerable module.

**Comment by mke2111** [comment]
## Project under review ⚪

**Review by mke2111 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED 🔴 ♻️ 

## Hey @prolajumokeoni, 👋🏽 
**This is a great project you have going on, well-done and thanks for implementing the changes requested earlier on**. 

### Well-done 💯 
    1. Most of the enumerable methods run just as expected.
    2. The linter checks run just fine as well.

**However, there are a few sections to improve before your project is approved.** 🔴  6 tests are failing currently.
<hr>

### REQUIRED

- [x] Check in-line comments for sections to improve.

### OPTIONAL

- [x] Add the cloning procedures along with the commands needed for one to set up your repo locally. You might as well add instructions to follow to test a given enumerable method.

#

### Kindly make all the requested changes and submit another CRR. 💯 

**Cheers 🍹 and Happy coding 💻** 

<hr>

Mention @mke2111 in this thread and I'll be glad to help. 😄 
### Best regards
Slack 🆘  Roy Mukuye
Twitter 🐦  [Roy Mukuye](https://twitter.com/Roymkenya)
LinkedIn 🔗  [Roy Mukuye](https://www.linkedin.com/in/roy-mukuye/)

File: enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        new_arr.push(yield item)
+      end
+    end
+    new_arr
+  end
+
+  # rubocop:disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+  def my_inject(acc_value = nil, value_two = nil)
+    if (!acc_value.nil? && value_twog.nil?) && (acc_value.is_a?(symbol) || acc_value.is_a?(string))
+      value_two = acc_value
```
Comment (mke2111) [inline enumerables.rb]: ##  `my_inject`

- [x] when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
```
Failure/Error: expect(array.my_inject(20, :*)).to eq array.inject(20, :*)
     
     NameError:
       undefined local variable or method `value_twog'
```
- You have a typo in there. It should be `value_two` and not `value_twog`.
- Giving this a quick fix will correct 3 tests. 😄 

- [x] when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
```
Failure/Error: expect(range.my_inject(:*)).to eq range.inject(:*)
     
     NameError:
       undefined local variable or method `symbol' for 5..50:Range



Failure/Error: expect(array.my_inject(:+)).to eq array.inject(:+)
     
     NameError:
       undefined local variable or method `symbol'
```
- Your method doesn't seem to recognize these operation symbols. Give this a quick fix as well.

- [x] when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
```
Failure/Error: actual = range.my_inject(4) { |prod, n| prod * n }
     
     NameError:
       undefined local variable or method `symbol' for 5..50:Range

```
- Give this section a fix as well.

Comment (prolajumokeoni) [reply enumerables.rb]: Thanks for the review @mke2111, I will implement the fix

**Comment by mimipeshy** [comment]
## STATUS: CHANGES REQUESTED 🔴 ♻️ 

Hello there @prolajumokeoni 👋🏾 👋🏾 , you have done an amazing job on this project. Most of your methods are working as expected. However, there is one method that is not working as expected. I will highlight it in the comment section.

#### TO IMPROVE

- [x]  Check in-line comments for sections to improve.

That's all. 😄 , kindy submit another review when you are done

Regards
Peris Ndanu
_If you have any questions, kindly leave a comment here and tag my name @mimipeshy and I will be glad to help_

File: enumerables.rb [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+  def my_each_with_index
+    # return to_enum(:my_each_with_index) unless block_given?
+    i = 0
+    while i < length
+      yield i, self[i]
+      i += 1
+    end
+    self
+  end
+
```
Comment (mimipeshy) [inline enumerables.rb]: - [x] Enumerables #my_each_with_index should return an enumerator if no block is given
- [x] Enumerables #my_each_with_index when a block is given when self is an array behaves like #each_with_index calls the given block once for each element in self
- [x] Enumerables #my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each element in self
  
     ` NameError: undefined local variable or method `length' for 5..50:Range`
- [x] Enumerables #my_each_with_index when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self

     ` NameError:
       undefined local variable or method `length' for 5..50:Range `

- [x] Enumerables #my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self

**Comment by mimipeshy** [comment]
## STATUS: APPROVED 🥳 👯 

Hello there @prolajumokeoni 👋🏾 👋🏾 , you have done an amazing job in this project. You should be proud of yourself 😄 All your methods are passing well. Linters are not forming any errors. Therefore I approve of this project, go ahead and merge.

Regards
Peris Ndanu
Happy Coding 💻 🖱️

**Pull request merged**